Address outstanding 2077 feedback

dunxen opened the pull request:
Resolves #2360.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7097,23 +7138,28 @@ where
 					}
-				} else { true };
-				if retain && chan.context.get_counterparty_node_id() != *counterparty_node_id {
+					retain
```
Yeah, that's true!

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3465,27 +3488,35 @@ where
+			let mut config = context.config();
 			config.apply(config_update);
-			if !channel.context.update_config(&config) {
+			if !context.update_config(&config) {
```
doh!

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7097,23 +7138,28 @@ where
-						});
-						true
+			peer_state.outbound_v1_channel_by_id.retain(|_, chan| {
+				if chan.context.get_counterparty_node_id() == *counterparty_node_id {
```
Ah, stale since 1507 lol. Will remove here and below.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7097,23 +7138,28 @@ where
+						node_id: chan.context.get_counterparty_node_id(),
+						msg: chan.get_channel_reestablish(&self.logger),
+					});
+				} else {
```
True!

However, I think I've come across something strange now in tests now where we get a mismatch between `chan.context.get_counterparty_node_id()` and the passed `counterparty_node_id` which sounds ridiculous (since we have channels per peer) so I'll check again with fresh eyes in the morning.

dunxen replied:
Oh yeah, we're iterating over all the peer states with the containing for-loop. Why would we be doing that if we're only interested in one counterparty_node_id? Tests also break without the body in the else, which deals with channels that don't correspond to our passed in `counterparty_node_id`.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7114,35 +7114,25 @@ where
+				log_error!(self.logger, "Force-closing channel {}", log_bytes!(chan.context.channel_id()));
+				self.issue_channel_close_events(&chan.context, ClosureReason::HolderForceClosed);
+				self.finish_force_close_channel(chan.context.force_shutdown(false));
+				false
```
Should we just `clear` after `iter`ating instead of using `retain`?

dunxen replied:
Yeah that’s more idiomatic since we never retain.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3494,27 +3494,35 @@ where
+				&mut channel.context
+			} else if let Some(channel) = peer_state.channel_by_id.get_mut(channel_id) {
+				let mut config = channel.context.config();
+				config.apply(config_update);
```
Does this actually work? `ChannelContext::config` returns a copy, not a reference. So the change is being applied to a local variable. We may be missing some test coverage if this is the case.

dunxen replied:
Uhhh yeah let me test that. Didn’t change any functionality (just moved) but could be missing that coverage!

jkczyz replied:
Oh, yeah, you're right. That's a strange interface but I guess it preexists the PR.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3494,27 +3494,35 @@ where
 				});
-			}
-		}
-		for channel_id in channel_ids {
```
:man_facepalming: Yeah, definitely didn't mean to remove this. Needs a test which I'll add.

dunxen commented:
> This is not the case -- we currently only have such disconnect logic on `ChannelReestablish` and `RevokeAndAck` messages. It can be extended to more messages though.

Thanks for pointing that out. I believe it does make sense to keep that disconnect logic in one place. And to me it also makes sense that we disconnect peers if they're not making progress on the prefunded parts of channel establishment. Even in the case where we already have funded channels with the peer it's probably best to still disconnect if they're not making progress establishing a new one, although I could be wrong.

dunxen commented:
Rebasing & adding disconnect logic.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7146,37 +7163,26 @@ where
+				self.issue_channel_close_events(&chan.context, ClosureReason::HolderForceClosed);
+				self.finish_force_close_channel(chan.context.force_shutdown(false));
+			}
+			// Drop all outbound channels which we have force-closed above.
```
Well it's not actually possible to have inbound pending channels here as that would mean we somehow received an `open_channel` when we were disconnected from our peer. Whenever we disconnect, we clear out all the pending inbound and outbound channels.

It won't hurt if we have code to handle the case which shouldn't happen or I can just leave some comments.

dunxen replied:
Ah then the disconnect cleanup isn’t done. Ok so we still need cleanup here for everything.

dunxen replied:
I can't imagine such a case, but will leave this commit up for a bit for comment while addressing your other feedback :)

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7146,37 +7163,26 @@ where
-					}
-				} else { true };
-				if retain && chan.context.get_counterparty_node_id() != *counterparty_node_id {
-					if let Some(msg) = chan.get_signed_channel_announcement(&self.node_signer, self.genesis_hash.clone(), self.best_block.read().unwrap().height(), &self.default_configuration) {
-						if let Ok(update_msg) = self.get_channel_update_for_broadcast(chan) {
-							pending_msg_events.push(events::MessageSendEvent::SendChannelAnnouncement {
-								node_id: *counterparty_node_id,
-								msg, update_msg,
-							});
-						}
-					}
```
I _think_ that block was a leftover from when channels were not stored per peer.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9457,6 +9457,33 @@ mod tests {
+			let per_peer_state = nodes[0].node.per_peer_state.read().unwrap();
+			let peer_state_mutex = per_peer_state.get(&nodes[1].node.get_our_node_id()).unwrap();
+			let mut peer_state = peer_state_mutex.lock().unwrap();
+			peer_state.is_connected = false;
```
Actually, we never persist pending channels so do we even need cleanup after a crash? I think the `peer_connected` cleanup for pending channels above may be redundant...

dunxen commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -10016,4 +10016,108 @@ fn test_disconnects_peer_awaiting_response_ticks() {
 	}
+
+	// V1 Channel Establishment
+	// Disconnects on pending channel timeout
```
Yeah. I'll try move out the helpers from this test to not duplicate them.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3542,27 +3542,34 @@ where
+			};
+			let mut config = context.config();
 			config.apply(config_update);
-			if !channel.context.update_config(&config) {
```
It's all below but just without the unwrap?

dunxen replied:
Yip. 🤦‍♂️ You’re right. Needs more tests for that too.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -6316,6 +6328,9 @@ impl<Signer: WriteableEcdsaChannelSigner> InboundV1Channel<Signer> {
 		self.context.inbound_awaiting_accept = false;
+		// If we're not making progress on a new pending channel we'll want to disconnect. So we mark the
+		// channel as awaiting response.
+		self.context.mark_awaiting_response();
```
Hmm. Is it? Manual acceptance also uses this method in `ChannelManager`.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -5792,6 +5798,10 @@ impl<Signer: WriteableEcdsaChannelSigner> OutboundV1Channel<Signer> {
 			panic!("Tried to send an open_channel for a channel that has already advanced");
 		}
 
+		// If we're not making progress on a new pending channel we'll want to disconnect. So we mark the
```
Ok will move forward with a total-funding-timeout for inbound/outbound chans. I'm imagining it's easiest to just also handle them with timer ticks and set to around 60 as limit then which should be +/- 1 hour.

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -590,6 +590,11 @@ pub(crate) const EXPIRE_PREV_CONFIG_TICKS: usize = 5;
+/// The number of ticks that may elapse while we're waiting for a pending outbound/inbound channel
+/// to be promoted to a [`Channel`] since the pending channel was created. A pending channel
+/// exceeding this age limit will be force-closed and purged from memory.
+pub(crate) const PENDING_CHANNEL_AGE_LIMIT_TICKS: usize = 60;
```
Yeah I’ve been sloppy with naming.

jkczyz replied:
nit: commit message says "ENDING_CHANNEL_AGE_LIMIT_TICKS"

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4414,6 +4415,28 @@ where
+					let force_close_expired_pending_channel = |chan_id: &[u8; 32], chan_context: &mut ChannelContext<<SP::Target as SignerProvider>::Signer>| {
+						log_error!(self.logger, "Force-closing pending outbound channel {} for not establishing in a timely manner", log_bytes!(&chan_id[..]));
+						self.issue_channel_close_events(&chan_context, ClosureReason::HolderForceClosed);
+						self.finish_force_close_channel(chan_context.force_shutdown(false));
```
Yes! I think we need to consolidate these calls :(

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -698,6 +698,18 @@ impl <Signer: ChannelSigner> PeerState<Signer> {
+	///       in `(outbound/inbound)_v1_channel_by_id`, we use the more general `channel_id` as V2
+	///       established channels will have a fixed `channel_id` already after the `accept_channel2`
+	///       message is sent/received.
+	fn has_pending_channel(&self, channel_id: &[u8; 32]) -> bool {
```
Should this go in the second commit?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2382,6 +2399,14 @@ where
 			let mut peer_state_lock = peer_state_mutex.lock().unwrap();
 			let peer_state = &mut *peer_state_lock;
+
+			if peer_state.has_pending_channel(&channel_id) {
```
Two or possibly three more map look-ups in `force_close_channel_with_peer` given this will be either a pending inbound or outbound channel and the non-pending map is checked first. So total of five look-ups for the inbound case.

FWIW, if there was one map with an enum, a single look-up could give back the entry to use by `force_close_channel_with_peer`. Maybe macro could help in the current design?

dunxen replied:
Yeah, single map with enum was considered. It's nice in cases like this, but I remember some friction elsewhere. Maybe something to consider again. I want to see how much worse it makes it in other areas again (and how big the commit would be to use an enum). Dropping the channel interface in the channel split PR probably makes it better now.

dunxen replied:
Ok so it's a bit involved to take a look at using an enum in a single map, but there are some nice advantages. I'll just continue with fixing up the multi-map approach and luckily it's something that we can change internally to channel monitor in a future PR.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5458,38 +5458,48 @@ where
-				hash_map::Entry::Occupied(mut chan_entry) => {
+			if let hash_map::Entry::Occupied(chan_entry) = peer_state.outbound_v1_channel_by_id.entry(msg.channel_id.clone()) {
+				log_error!(self.logger, "Force-closing channel {}", log_bytes!(&msg.channel_id[..]));
+				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::HolderForceClosed);
```
I think I’ll just introduce a new variant called “UnfundedChannelExpired” or whatever.

dunxen replied:
Oh wait. Got mixed up with the threads 👍

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2392,56 +2380,58 @@ where
+			// If we reach this point, it means that the channel_id either refers to an unfunded channel or
+			// it does not exists for this peer. Either way, we can attempt to force-close it.
+			//
+			// An appropriate error will be returned forn non-existence of the channel if that's the case.
```
s/forn/for

jkczyz commented on lightning/src/events/mod.rs:
```diff
@@ -164,7 +164,10 @@ pub enum ClosureReason {
 	/// [`ChannelManager`]: crate::ln::channelmanager::ChannelManager
-	OutdatedChannelManager
+	OutdatedChannelManager,
+	/// The counterparty requested a cooperative close of a channel that had not established
+	/// a funding transaction yet. The channel has been immediately closed.
```
A little wordy. Maybe: "had not been funded yet"

jkczyz commented on lightning/src/events/mod.rs:
```diff
@@ -164,7 +164,10 @@ pub enum ClosureReason {
+	OutdatedChannelManager,
+	/// The counterparty requested a cooperative close of a channel that had not established
+	/// a funding transaction yet. The channel has been immediately closed.
+	CounterpartyCoopClosedPrefundedChannel,
```
Hmm... "prefunded" makes it sound like it was funded in advance. Shouldn't we say "unfunded"? Might be another occurrence of this predating the PR.

dunxen replied:
Yeah exactly the reason I’ve started changing and avoiding “refunded”, but this was accidental, not preexisting 🤦‍♂️

dunxen commented:
Rebased and squashed with:

* `pending_context` -> `unfunded_context`
* A move of the import for `UnfundedChannelContext`
* @jkczyz's latest feedback resolved

```range-diff
 1:  c93b709d =  1:  3294c678 Add missing unfunded channel maps checks in `ChannelManager`
 2:  b10127c1 <  -:  -------- Fallback `close_channel_internal` to force close unfunded channels
 3:  42d78374 <  -:  -------- f avoid excess map lookups & lockorder assertions
 -:  -------- >  2:  0d6e3033 Fallback `close_channel_internal` to force close unfunded channels
 4:  a68d556e !  3:  ea515daf Consider all channel maps in `update_partial_channel_config`
    @@ lightning/src/ln/channelmanager.rs: where
     +			} else if let Some(channel) = peer_state.outbound_v1_channel_by_id.get_mut(channel_id) {
     +				&mut channel.context
     +			} else {
    ++				// This should not be reachable as we've already checked for non-existence in the previous channel_id loop.
    ++				debug_assert!(false);
     +				return Err(APIError::ChannelUnavailable {
    -+					err: format!("Channel with ID {} was not found for the passed counterparty_node_id {}", log_bytes!(*channel_id), counterparty_node_id),
    ++					err: format!(
    ++						"Channel with ID {} for passed counterparty_node_id {} disappeared after we confirmed its existence - this should not be reachable!",
    ++						log_bytes!(*channel_id), counterparty_node_id),
      				});
     -			}
     +			};
 5:  ecb44f40 <  -:  -------- f debug_assert false unreachable non-existence
 6:  c9ca69e0 !  4:  b4881d99 Force close pending channels in `internal_shutdown`
    @@ Metadata
      ## Commit message ##
         Force close pending channels in `internal_shutdown`
     
    + ## lightning/src/events/mod.rs ##
    +@@ lightning/src/events/mod.rs: impl_writeable_tlv_based_enum_upgradable!(PathFailure,
    + );
    + 
    + #[derive(Clone, Debug, PartialEq, Eq)]
    +-/// The reason the channel was closed. See individual variants more details.
    ++/// The reason the channel was closed. See individual variants for more details.
    + pub enum ClosureReason {
    + 	/// Closure generated from receiving a peer error message.
    + 	///
    +@@ lightning/src/events/mod.rs: pub enum ClosureReason {
    + 	///
    + 	/// [`ChannelMonitor`]: crate::chain::channelmonitor::ChannelMonitor
    + 	/// [`ChannelManager`]: crate::ln::channelmanager::ChannelManager
    +-	OutdatedChannelManager
    ++	OutdatedChannelManager,
    ++	/// The counterparty requested a cooperative close of a channel that had not been funded yet.
    ++	/// The channel has been immediately closed.
    ++	CounterpartyCoopClosedUnfundedChannel,
    + }
    + 
    + impl core::fmt::Display for ClosureReason {
    +@@ lightning/src/events/mod.rs: impl core::fmt::Display for ClosureReason {
    + 			},
    + 			ClosureReason::DisconnectedPeer => f.write_str("the peer disconnected prior to the channel being funded"),
    + 			ClosureReason::OutdatedChannelManager => f.write_str("the ChannelManager read from disk was stale compared to ChannelMonitor(s)"),
    ++			ClosureReason::CounterpartyCoopClosedUnfundedChannel => f.write_str("the peer requested the prefunded channel be closed"),
    + 		}
    + 	}
    + }
    +@@ lightning/src/events/mod.rs: impl_writeable_tlv_based_enum_upgradable!(ClosureReason,
    + 	(8, ProcessingError) => { (1, err, required) },
    + 	(10, DisconnectedPeer) => {},
    + 	(12, OutdatedChannelManager) => {},
    ++	(13, CounterpartyCoopClosedUnfundedChannel) => {},
    + );
    + 
    + /// Intended destination of a failed HTLC as indicated in [`Event::HTLCHandlingFailed`].
    +
      ## lightning/src/ln/channelmanager.rs ##
     @@ lightning/src/ln/channelmanager.rs: where
      				})?;
    @@ lightning/src/ln/channelmanager.rs: where
     -					let (shutdown, monitor_update_opt, htlcs) = try_chan_entry!(self,
     -						chan_entry.get_mut().shutdown(&self.signer_provider, &peer_state.latest_features, &msg), chan_entry);
     -					dropped_htlcs = htlcs;
    ++			// TODO(dunxen): Fix this duplication when we switch to a single map with enums as per
    ++			// https://github.com/lightningdevkit/rust-lightning/issues/2422
     +			if let hash_map::Entry::Occupied(chan_entry) = peer_state.outbound_v1_channel_by_id.entry(msg.channel_id.clone()) {
    -+				log_error!(self.logger, "Force-closing channel {}", log_bytes!(&msg.channel_id[..]));
    -+				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::HolderForceClosed);
    ++				log_error!(self.logger, "Immediately closing prefunded channel {} as peer asked to cooperatively shut it down (which is unnecessary)", log_bytes!(&msg.channel_id[..]));
    ++				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::CounterpartyCoopClosedUnfundedChannel);
     +				let mut chan = remove_channel!(self, chan_entry);
     +				self.finish_force_close_channel(chan.context.force_shutdown(false));
     +				return Ok(());
     +			} else if let hash_map::Entry::Occupied(chan_entry) = peer_state.inbound_v1_channel_by_id.entry(msg.channel_id.clone()) {
    -+				log_error!(self.logger, "Force-closing channel {}", log_bytes!(&msg.channel_id[..]));
    -+				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::HolderForceClosed);
    ++				log_error!(self.logger, "Immediately closing prefunded channel {} as peer asked to cooperatively shut it down (which is unnecessary)", log_bytes!(&msg.channel_id[..]));
    ++				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::CounterpartyCoopClosedUnfundedChannel);
     +				let mut chan = remove_channel!(self, chan_entry);
     +				self.finish_force_close_channel(chan.context.force_shutdown(false));
     +				return Ok(());
 7:  754c5a53 <  -:  -------- f add ClosureReason::CounterpartyCoopClosedPrefundedChannel
 8:  5996af18 =  5:  250690af Remove redundant 'outbound' wording from methods
 9:  c446ae5c !  6:  107f42f3 Close and remove unfunded inbound/outbound channels that are older than an hour
    @@ lightning/src/ln/channel.rs: impl<Signer: WriteableEcdsaChannelSigner> Channel<S
      /// A not-yet-funded outbound (from holder) channel using V1 channel establishment.
      pub(super) struct OutboundV1Channel<Signer: ChannelSigner> {
      	pub context: ChannelContext<Signer>,
    -+	pub pending_context: UnfundedChannelContext,
    ++	pub unfunded_context: UnfundedChannelContext,
      }
      
      impl<Signer: WriteableEcdsaChannelSigner> OutboundV1Channel<Signer> {
    @@ lightning/src/ln/channel.rs: impl<Signer: WriteableEcdsaChannelSigner> OutboundV
      				blocked_monitor_updates: Vec::new(),
     -			}
     +			},
    -+			pending_context: UnfundedChannelContext { unfunded_channel_age_ticks: 0 }
    ++			unfunded_context: UnfundedChannelContext { unfunded_channel_age_ticks: 0 }
      		})
      	}
      
    @@ lightning/src/ln/channel.rs: impl<Signer: WriteableEcdsaChannelSigner> OutboundV
      /// A not-yet-funded inbound (from counterparty) channel using V1 channel establishment.
      pub(super) struct InboundV1Channel<Signer: ChannelSigner> {
      	pub context: ChannelContext<Signer>,
    -+	pub pending_context: UnfundedChannelContext,
    ++	pub unfunded_context: UnfundedChannelContext,
      }
      
      impl<Signer: WriteableEcdsaChannelSigner> InboundV1Channel<Signer> {
    @@ lightning/src/ln/channel.rs: impl<Signer: WriteableEcdsaChannelSigner> InboundV1
      				blocked_monitor_updates: Vec::new(),
     -			}
     +			},
    -+			pending_context: UnfundedChannelContext { unfunded_channel_age_ticks: 0 }
    ++			unfunded_context: UnfundedChannelContext { unfunded_channel_age_ticks: 0 }
      		};
      
      		Ok(chan)
     
      ## lightning/src/ln/channelmanager.rs ##
    +@@ lightning/src/ln/channelmanager.rs: use crate::events::{Event, EventHandler, EventsProvider, MessageSendEvent, Messa
    + // Since this struct is returned in `list_channels` methods, expose it here in case users want to
    + // construct one themselves.
    + use crate::ln::{inbound_payment, PaymentHash, PaymentPreimage, PaymentSecret};
    +-use crate::ln::channel::{Channel, ChannelContext, ChannelError, ChannelUpdateStatus, ShutdownResult, UpdateFulfillCommitFetch, OutboundV1Channel, InboundV1Channel};
    ++use crate::ln::channel::{Channel, ChannelContext, ChannelError, ChannelUpdateStatus, ShutdownResult, UnfundedChannelContext, UpdateFulfillCommitFetch, OutboundV1Channel, InboundV1Channel};
    + use crate::ln::features::{ChannelFeatures, ChannelTypeFeatures, InitFeatures, NodeFeatures};
    + #[cfg(any(feature = "_test_utils", test))]
    + use crate::ln::features::Bolt11InvoiceFeatures;
     @@ lightning/src/ln/channelmanager.rs: where
      	///  * Expiring a channel's previous [`ChannelConfig`] if necessary to only allow forwarding HTLCs
      	///    with the current [`ChannelConfig`].
    @@ lightning/src/ln/channelmanager.rs: where
      						true
      					});
     +
    -+					let force_close_expired_unfunded_channel = |chan_id: &[u8; 32], chan_context: &mut ChannelContext<<SP::Target as SignerProvider>::Signer>| {
    -+						log_error!(self.logger, "Force-closing pending outbound channel {} for not establishing in a timely manner", log_bytes!(&chan_id[..]));
    -+						self.issue_channel_close_events(&chan_context, ClosureReason::HolderForceClosed);
    -+						self.finish_force_close_channel(chan_context.force_shutdown(false));
    -+						false
    -+					};
    -+					peer_state.outbound_v1_channel_by_id.retain(|chan_id, chan| {
    -+						if chan.pending_context.should_expire_unfunded_channel() {
    -+							force_close_expired_unfunded_channel(chan_id, &mut chan.context)
    ++					let process_unfunded_channel_tick = |
    ++						chan_id: &[u8; 32],
    ++						chan_context: &mut ChannelContext<<SP::Target as SignerProvider>::Signer>,
    ++						unfunded_chan_context: &mut UnfundedChannelContext,
    ++					| {
    ++						chan_context.maybe_expire_prev_config();
    ++						if unfunded_chan_context.should_expire_unfunded_channel() {
    ++							log_error!(self.logger, "Force-closing pending outbound channel {} for not establishing in a timely manner", log_bytes!(&chan_id[..]));
    ++							update_maps_on_chan_removal!(self, &chan_context);
    ++							self.issue_channel_close_events(&chan_context, ClosureReason::HolderForceClosed);
    ++							self.finish_force_close_channel(chan_context.force_shutdown(false));
    ++							false
     +						} else {
     +							true
     +						}
    -+					});
    -+					peer_state.inbound_v1_channel_by_id.retain(|chan_id, chan| {
    -+						if chan.pending_context.should_expire_unfunded_channel() {
    -+							force_close_expired_unfunded_channel(chan_id, &mut chan.context)
    -+						} else {
    -+							true
    -+						}
    -+					});
    ++					};
    ++					peer_state.outbound_v1_channel_by_id.retain(|chan_id, chan| process_unfunded_channel_tick(chan_id, &mut chan.context, &mut chan.unfunded_context));
    ++					peer_state.inbound_v1_channel_by_id.retain(|chan_id, chan| process_unfunded_channel_tick(chan_id, &mut chan.context, &mut chan.unfunded_context));
     +
      					if peer_state.ok_to_remove(true) {
      						pending_peers_awaiting_removal.push(counterparty_node_id);
10:  1c8ef1bf <  -:  -------- f missing calls in tick
```

jkczyz commented:
> Rebased and squashed with:
> 
> * `pending_context` -> `unfunded_context`
> * A move of the import for `UnfundedChannelContext`
> * @jkczyz's latest feedback resolved
> 
> ```
>  1:  c93b709d =  1:  3294c678 Add missing unfunded channel maps checks in `ChannelManager`
>  2:  b10127c1 <  -:  -------- Fallback `close_channel_internal` to force close unfunded channels
>  3:  42d78374 <  -:  -------- f avoid excess map lookups & lockorder assertions
>  -:  -------- >  2:  0d6e3033 Fallback `close_channel_internal` to force close unfunded channels
>  4:  a68d556e !  3:  ea515daf Consider all channel maps in `update_partial_channel_config`
>     @@ lightning/src/ln/channelmanager.rs: where
>      +			} else if let Some(channel) = peer_state.outbound_v1_channel_by_id.get_mut(channel_id) {
>      +				&mut channel.context
>      +			} else {
>     ++				// This should not be reachable as we've already checked for non-existence in the previous channel_id loop.
>     ++				debug_assert!(false);
>      +				return Err(APIError::ChannelUnavailable {
>     -+					err: format!("Channel with ID {} was not found for the passed counterparty_node_id {}", log_bytes!(*channel_id), counterparty_node_id),
>     ++					err: format!(
>     ++						"Channel with ID {} for passed counterparty_node_id {} disappeared after we confirmed its existence - this should not be reachable!",
>     ++						log_bytes!(*channel_id), counterparty_node_id),
>       				});
>      -			}
>      +			};
>  5:  ecb44f40 <  -:  -------- f debug_assert false unreachable non-existence
>  6:  c9ca69e0 !  4:  b4881d99 Force close pending channels in `internal_shutdown`
>     @@ Metadata
>       ## Commit message ##
>          Force close pending channels in `internal_shutdown`
>      
>     + ## lightning/src/events/mod.rs ##
>     +@@ lightning/src/events/mod.rs: impl_writeable_tlv_based_enum_upgradable!(PathFailure,
>     + );
>     + 
>     + #[derive(Clone, Debug, PartialEq, Eq)]
>     +-/// The reason the channel was closed. See individual variants more details.
>     ++/// The reason the channel was closed. See individual variants for more details.
>     + pub enum ClosureReason {
>     + 	/// Closure generated from receiving a peer error message.
>     + 	///
>     +@@ lightning/src/events/mod.rs: pub enum ClosureReason {
>     + 	///
>     + 	/// [`ChannelMonitor`]: crate::chain::channelmonitor::ChannelMonitor
>     + 	/// [`ChannelManager`]: crate::ln::channelmanager::ChannelManager
>     +-	OutdatedChannelManager
>     ++	OutdatedChannelManager,
>     ++	/// The counterparty requested a cooperative close of a channel that had not been funded yet.
>     ++	/// The channel has been immediately closed.
>     ++	CounterpartyCoopClosedUnfundedChannel,
>     + }
>     + 
>     + impl core::fmt::Display for ClosureReason {
>     +@@ lightning/src/events/mod.rs: impl core::fmt::Display for ClosureReason {
>     + 			},
>     + 			ClosureReason::DisconnectedPeer => f.write_str("the peer disconnected prior to the channel being funded"),
>     + 			ClosureReason::OutdatedChannelManager => f.write_str("the ChannelManager read from disk was stale compared to ChannelMonitor(s)"),
>     ++			ClosureReason::CounterpartyCoopClosedUnfundedChannel => f.write_str("the peer requested the prefunded channel be closed"),
>     + 		}
>     + 	}
>     + }
>     +@@ lightning/src/events/mod.rs: impl_writeable_tlv_based_enum_upgradable!(ClosureReason,
>     + 	(8, ProcessingError) => { (1, err, required) },
>     + 	(10, DisconnectedPeer) => {},
>     + 	(12, OutdatedChannelManager) => {},
>     ++	(13, CounterpartyCoopClosedUnfundedChannel) => {},
>     + );
>     + 
>     + /// Intended destination of a failed HTLC as indicated in [`Event::HTLCHandlingFailed`].
>     +
>       ## lightning/src/ln/channelmanager.rs ##
>      @@ lightning/src/ln/channelmanager.rs: where
>       				})?;
>     @@ lightning/src/ln/channelmanager.rs: where
>      -					let (shutdown, monitor_update_opt, htlcs) = try_chan_entry!(self,
>      -						chan_entry.get_mut().shutdown(&self.signer_provider, &peer_state.latest_features, &msg), chan_entry);
>      -					dropped_htlcs = htlcs;
>     ++			// TODO(dunxen): Fix this duplication when we switch to a single map with enums as per
>     ++			// https://github.com/lightningdevkit/rust-lightning/issues/2422
>      +			if let hash_map::Entry::Occupied(chan_entry) = peer_state.outbound_v1_channel_by_id.entry(msg.channel_id.clone()) {
>     -+				log_error!(self.logger, "Force-closing channel {}", log_bytes!(&msg.channel_id[..]));
>     -+				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::HolderForceClosed);
>     ++				log_error!(self.logger, "Immediately closing prefunded channel {} as peer asked to cooperatively shut it down (which is unnecessary)", log_bytes!(&msg.channel_id[..]));
>     ++				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::CounterpartyCoopClosedUnfundedChannel);
>      +				let mut chan = remove_channel!(self, chan_entry);
>      +				self.finish_force_close_channel(chan.context.force_shutdown(false));
>      +				return Ok(());
>      +			} else if let hash_map::Entry::Occupied(chan_entry) = peer_state.inbound_v1_channel_by_id.entry(msg.channel_id.clone()) {
>     -+				log_error!(self.logger, "Force-closing channel {}", log_bytes!(&msg.channel_id[..]));
>     -+				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::HolderForceClosed);
>     ++				log_error!(self.logger, "Immediately closing prefunded channel {} as peer asked to cooperatively shut it down (which is unnecessary)", log_bytes!(&msg.channel_id[..]));
>     ++				self.issue_channel_close_events(&chan_entry.get().context, ClosureReason::CounterpartyCoopClosedUnfundedChannel);
>      +				let mut chan = remove_channel!(self, chan_entry);
>      +				self.finish_force_close_channel(chan.context.force_shutdown(false));
>      +				return Ok(());
>  7:  754c5a53 <  -:  -------- f add ClosureReason::CounterpartyCoopClosedPrefundedChannel
>  8:  5996af18 =  5:  250690af Remove redundant 'outbound' wording from methods
>  9:  c446ae5c !  6:  107f42f3 Close and remove unfunded inbound/outbound channels that are older than an hour
>     @@ lightning/src/ln/channel.rs: impl<Signer: WriteableEcdsaChannelSigner> Channel<S
>       /// A not-yet-funded outbound (from holder) channel using V1 channel establishment.
>       pub(super) struct OutboundV1Channel<Signer: ChannelSigner> {
>       	pub context: ChannelContext<Signer>,
>     -+	pub pending_context: UnfundedChannelContext,
>     ++	pub unfunded_context: UnfundedChannelContext,
>       }
>       
>       impl<Signer: WriteableEcdsaChannelSigner> OutboundV1Channel<Signer> {
>     @@ lightning/src/ln/channel.rs: impl<Signer: WriteableEcdsaChannelSigner> OutboundV
>       				blocked_monitor_updates: Vec::new(),
>      -			}
>      +			},
>     -+			pending_context: UnfundedChannelContext { unfunded_channel_age_ticks: 0 }
>     ++			unfunded_context: UnfundedChannelContext { unfunded_channel_age_ticks: 0 }
>       		})
>       	}
>       
>     @@ lightning/src/ln/channel.rs: impl<Signer: WriteableEcdsaChannelSigner> OutboundV
>       /// A not-yet-funded inbound (from counterparty) channel using V1 channel establishment.
>       pub(super) struct InboundV1Channel<Signer: ChannelSigner> {
>       	pub context: ChannelContext<Signer>,
>     -+	pub pending_context: UnfundedChannelContext,
>     ++	pub unfunded_context: UnfundedChannelContext,
>       }
>       
>       impl<Signer: WriteableEcdsaChannelSigner> InboundV1Channel<Signer> {
>     @@ lightning/src/ln/channel.rs: impl<Signer: WriteableEcdsaChannelSigner> InboundV1
>       				blocked_monitor_updates: Vec::new(),
>      -			}
>      +			},
>     -+			pending_context: UnfundedChannelContext { unfunded_channel_age_ticks: 0 }
>     ++			unfunded_context: UnfundedChannelContext { unfunded_channel_age_ticks: 0 }
>       		};
>       
>       		Ok(chan)
>      
>       ## lightning/src/ln/channelmanager.rs ##
>     +@@ lightning/src/ln/channelmanager.rs: use crate::events::{Event, EventHandler, EventsProvider, MessageSendEvent, Messa
>     + // Since this struct is returned in `list_channels` methods, expose it here in case users want to
>     + // construct one themselves.
>     + use crate::ln::{inbound_payment, PaymentHash, PaymentPreimage, PaymentSecret};
>     +-use crate::ln::channel::{Channel, ChannelContext, ChannelError, ChannelUpdateStatus, ShutdownResult, UpdateFulfillCommitFetch, OutboundV1Channel, InboundV1Channel};
>     ++use crate::ln::channel::{Channel, ChannelContext, ChannelError, ChannelUpdateStatus, ShutdownResult, UnfundedChannelContext, UpdateFulfillCommitFetch, OutboundV1Channel, InboundV1Channel};
>     + use crate::ln::features::{ChannelFeatures, ChannelTypeFeatures, InitFeatures, NodeFeatures};
>     + #[cfg(any(feature = "_test_utils", test))]
>     + use crate::ln::features::Bolt11InvoiceFeatures;
>      @@ lightning/src/ln/channelmanager.rs: where
>       	///  * Expiring a channel's previous [`ChannelConfig`] if necessary to only allow forwarding HTLCs
>       	///    with the current [`ChannelConfig`].
>     @@ lightning/src/ln/channelmanager.rs: where
>       						true
>       					});
>      +
>     -+					let force_close_expired_unfunded_channel = |chan_id: &[u8; 32], chan_context: &mut ChannelContext<<SP::Target as SignerProvider>::Signer>| {
>     -+						log_error!(self.logger, "Force-closing pending outbound channel {} for not establishing in a timely manner", log_bytes!(&chan_id[..]));
>     -+						self.issue_channel_close_events(&chan_context, ClosureReason::HolderForceClosed);
>     -+						self.finish_force_close_channel(chan_context.force_shutdown(false));
>     -+						false
>     -+					};
>     -+					peer_state.outbound_v1_channel_by_id.retain(|chan_id, chan| {
>     -+						if chan.pending_context.should_expire_unfunded_channel() {
>     -+							force_close_expired_unfunded_channel(chan_id, &mut chan.context)
>     ++					let process_unfunded_channel_tick = |
>     ++						chan_id: &[u8; 32],
>     ++						chan_context: &mut ChannelContext<<SP::Target as SignerProvider>::Signer>,
>     ++						unfunded_chan_context: &mut UnfundedChannelContext,
>     ++					| {
>     ++						chan_context.maybe_expire_prev_config();
>     ++						if unfunded_chan_context.should_expire_unfunded_channel() {
>     ++							log_error!(self.logger, "Force-closing pending outbound channel {} for not establishing in a timely manner", log_bytes!(&chan_id[..]));
>     ++							update_maps_on_chan_removal!(self, &chan_context);
>     ++							self.issue_channel_close_events(&chan_context, ClosureReason::HolderForceClosed);
>     ++							self.finish_force_close_channel(chan_context.force_shutdown(false));
>     ++							false
>      +						} else {
>      +							true
>      +						}
>     -+					});
>     -+					peer_state.inbound_v1_channel_by_id.retain(|chan_id, chan| {
>     -+						if chan.pending_context.should_expire_unfunded_channel() {
>     -+							force_close_expired_unfunded_channel(chan_id, &mut chan.context)
>     -+						} else {
>     -+							true
>     -+						}
>     -+					});
>     ++					};
>     ++					peer_state.outbound_v1_channel_by_id.retain(|chan_id, chan| process_unfunded_channel_tick(chan_id, &mut chan.context, &mut chan.unfunded_context));
>     ++					peer_state.inbound_v1_channel_by_id.retain(|chan_id, chan| process_unfunded_channel_tick(chan_id, &mut chan.context, &mut chan.unfunded_context));
>      +
>       					if peer_state.ok_to_remove(true) {
>       						pending_peers_awaiting_removal.push(counterparty_node_id);
> 10:  1c8ef1bf <  -:  -------- f missing calls in tick
> ```

The `Display` string and error logging are still using "prefunded".

dunxen commented:
Changes some occurrences of "prefunded" to "unfunded" in first commit.

jkczyz commented:
> Just changed some occurrences of "prefunded" to "unfunded" in first commit.
> 
> EDIT: Ugh. Missed the actual ones Jeff pointed out. Fixed now.

FWIW, regarding the commit message and change, I'm indifferent to use of "pending". I primarily thought "prefunded" wasn't accurate.

dunxen commented:
> > Just changed some occurrences of "prefunded" to "unfunded" in first commit.
> > EDIT: Ugh. Missed the actual ones Jeff pointed out. Fixed now.
> 
> FWIW, regarding the commit message and change, I'm indifferent to use of "pending". I primarily thought "prefunded" wasn't accurate.

@wpaulino pointed out that pending is a little ambiguous, which I agree with as it might refer to "pending confirmation". So "unfunded" is the best short name we're going to get :)

dunxen commented on lightning/src/ln/channel.rs:
```diff
@@ -590,6 +590,11 @@ pub(crate) const EXPIRE_PREV_CONFIG_TICKS: usize = 5;
 /// See [`ChannelContext::sent_message_awaiting_response`] for more information.
 pub(crate) const DISCONNECT_PEER_AWAITING_RESPONSE_TICKS: usize = 2;
 
+/// The number of ticks that may elapse while we're waiting for a pending outbound/inbound channel
+/// to be promoted to a [`Channel`] since the unfunded channel was created. An unfunded channel
+/// exceeding this age limit will be force-closed and purged from memory.
```
I can fix in the enum map follow up after 116 release if this PR is holding that up.

dunxen commented:
Rebased and changed the one occurrence of "pending" to "unfunded" that Jeff pointed out.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7226,37 +7295,20 @@ where
-			peer_state.channel_by_id.retain(|_, chan| {
-				let retain = if chan.context.get_counterparty_node_id() == *counterparty_node_id {
-					if !chan.context.have_received_message() {
-						// If we created this (outbound) channel while we were disconnected from the
```
I can’t think of a situation where inbound / outbound sets wouldn’t be empty upon reconnect. We clear them on disconnect and they’re never persisted so aren’t recoverable after a crash. Or I might be missing a case.
